src/estimation/include/iDynTree/Estimation/AttitudeMahonyFilter.h

```diff
@@ -0,0 +1,169 @@
+ * \f$ q_{k+1} = q_{k} + \Delta t \frac{1}{2}q_{k} \circ \begin{bmatrix} 0 \\ \Omega_y_{k+1} - b_k + K_p \omega_{mes_{k+1}}\end{bmatrix}\f$
+ * \f$ \Omega_{k+1} = \Omega_y_{k+1} - b_k \f$
+ * \f$ b_{k+1} = b_k - K_i \Delta t \frac{1}{2} \omega_{mes_{k+1}} \f$
+ *
+ * The updateFilterWithMeasurements() uses the recent IMU measurements to compute the term \f$ \omega_{mes} \f$ which gives the vectorial from accelerometer and magnetometer measurements
+ * \f$ \omega_{mes} = -(\Sigma{n}{i=1} \frac{k_i}{2} (v_i \hat{v}_i^T - \hat{v}_i v_i^T) )^{\vee} \f$
+ * where \f$ v_i \f$ is the normalized accelerometer or magnetometer measurement,
+ * \f$ \hat{v_i} \f$ is the vector obtained from the orientation estimated combined with gravity direction or absolute magnetic field direction, for e.g, \f$ \hat{v_acc} = {^w}R_b^T e_3 \f$
+ * and \f$ k_i \f$ is the confidence weight on the i-th measurement. In our case, i = 1 or 2.
+ *
+ * The usage of the attitude estimator can be as follows,
+ * - After instantiation, the parameters of the filter can be set using the individual parameter methods or the struct method.
+ * - The filter state can be initialized by calling the setInternalState() method
+ * - Once initialized, the following filter methods can be run in a loop to get the orientation estimates,
+ *     - updateFilterWithMeasurements() method to pass the recent measurements to the filter
+ *     - propagateStates() method to propagate the states through the system dynamics and correcting using the updated measurements
+ *     - getInternalState() or getOrientationEstimate*() methods to get the entire state estimate or only the attitude estimated in desired representation
+ */
+class AttitudeMahonyFilter : public IAttitudeEstimator
+{
+public:
+    AttitudeMahonyFilter();
+
+    /**
+     * @brief set flag to use magnetometer measurements
+     * @param[in] flag enable/disable magnetometer measurements
+     */
+    void useMagnetoMeterMeasurements(bool flag);
+
+    /**
+     * @brief set the confidence weights on magenetometer measurements, if used
+     * @param[in] condifence can take values between \f$ [0, 1] \f$
+     */
+    void setConfidenceForMagnetometerMeasurements(double confidence);
+
+    /**
+     * @brief set the Kp gain
+     * @param[in] kp gain
+     */
+    void setGainkp(double kp);
+
+    /**
+     * @brief set the Ki gain
+     * @param[in] ki gain
+     */
+    void setGainki(double ki);
+
+    /**
+     * @brief set discretization time step in seconds
+     * @param[in] timestepInSeconds time step
+     */
+    void setTimeStepInSeconds(double timestepInSeconds);
+
+    /**
+     * @brief Set the gravity direction assumed by the filter (for computing orientation vectorial from accelerometer)
+     * @param[in] gravity_dir gravity direction
+     */
+    void setGravityDirection(const iDynTree::Direction& gravity_dir);
```

Comment:
Do you think it is handy to have the single-parameter methods if you already have the `setParameters`/`getParameters` call?

Reply:
Single parameter methods come really handy for RPC calls. I was actually considering of having only single parameter methods instead of the entire parameter methods. Since, the entire parameter methods complicate a bit the process (for instance in EKF), since it also involves useMagnetometerMeasurement() to be called. 
Why this is troublesome ? 
The prevention of memory leaks in the EKF is mainly done by the initializeFilter() method. Before calling this method, the EKF needs to know the dimensions of the states, inputs and outputs. IN cas of toggling the use_magnetometer_measurement flag to true, this means that the output dimensions has changed and the filter buffers has to be resized and reinitialized.

Arbitrarily calling setParams() without making any change to the use_magnetometer_measurement flag, makes it call the initializeFilter() forcefully, when it is not required. This adds a complexity to the design. 

At the moment, all this is being handled in a way which I don't know is elegant or graceful. But in my tests, it has not resulted in a leak. However, it is a crime for the author of the code to test the code and say it is working fine.

Reply:
> Why this is troublesome ?

Maintaining the per-parameter methods are just a slightly increase in maintenance burden, but if in your opinion the gain in usability is worth it, I think it make sense.